proving.go

```diff
@@ -0,0 +1,260 @@
+package merkle
+
+import (
+	"errors"
+	"math"
+)
+
+const NodeSize = 32
+
+type NodeReader interface {
+	Seek(index uint64) error
+	ReadNext() ([]byte, error)
+	Width() uint64
+}
+
+func GenerateProof(
+	provenLeafIndices []uint64,
+	readers map[uint]NodeReader,
+	hash HashFunc,
+) ([][]byte, error) {
+
+	var currentPos, subtreeStart position
+	var width uint64
+	var proof, additionalProof [][]byte
+
+	provenLeafIndexIt := &positionsIterator{s: provenLeafIndices}
+	skipPositions := &positionsStack{}
+	rootHeight := rootHeightFromWidth(readers[0].Width())
+
+	cache, err := NewTreeCache(readers, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	for ; ; currentPos = currentPos.parent() { // Process proven leaves:
```

Comment:
don't understand why you need to "increment" `currentPos`.. Anyway you assign it with a new value when calling `subtreeDefinition`

Reply:
Good point, it's only needed in the inner loop. Fixing.